### PR TITLE
Add documentation for `FileService` & `FileSystemProvider`

### PR DESCRIPTION
**Pull request by tortmayr**
Signed-off-by: Tobias Ortmayr <tortmayr@eclipsesource.com>
Contributed on behalf of STMicroelectronics

### CODE REVIEW

File: packages/filesystem/src/browser/file-service.ts [hunk 1]
```diff
@@ -69,10 +69,10 @@ import { FileSystemWatcherErrorHandler } from './filesystem-watcher-error-handle
 
 export interface FileOperationParticipant {
 
-	/**
-	 * Participate in a file operation of a working copy. Allows to
-	 * change the working copy before it is being saved to disk.
-	 */
+    /**
```
Comment (koegel) [inline packages/filesystem/src/browser/file-service.ts:72]: Is all this whitespace change really intended?

File: packages/filesystem/src/browser/file-service.ts [hunk 2]
```diff
@@ -151,60 +151,94 @@ export interface UpdateTextFileOptions extends WriteEncodingOptions, WriteFileOp
...
+ * Represents the `FileSystemProviderRegistration` event.
+ * This event is fired by the {@link FileService} if a {@link FileSystemProvider} is
+ * registered to or unregistered from the service.
+ */
 export interface FileSystemProviderRegistrationEvent {
+    /** `True` if a new provider has been registered, `false` if a provider has been unregistered. */
     added: boolean;
+    /** The (uri) scheme for which the provider was (previously) registered */
     scheme: string;
+    /** The affected file system provider for which this event was fired. */
```
Comment (koegel) [inline packages/filesystem/src/browser/file-service.ts:219]: Why is it optional?

File: packages/filesystem/src/browser/file-service.ts [hunk 3]
```diff
@@ -225,6 +259,12 @@ export class TextFileOperationError extends FileOperationError {
 
 }
 
+/**
+ * The {@link FileService} is the common frontend component responsible for all interactions with file systems.
+ * This service manages all registered {@link FileSystemProvider}. In a nutshell, the {@link FileService} acts as a facade for all registered
```
Comment (koegel) [inline packages/filesystem/src/browser/file-service.ts:264]:
```suggestion
 * It manages all registered {@link FileSystemProvider}s and
```

File: packages/filesystem/src/browser/file-service.ts [hunk 4]
```diff
@@ -225,6 +259,12 @@ export class TextFileOperationError extends FileOperationError {
 
 }
 
+/**
+ * The {@link FileService} is the common frontend component responsible for all interactions with file systems.
```
Comment (koegel) [inline packages/filesystem/src/browser/file-service.ts:263]:
```suggestion
 * The {@link FileService} is the common facade responsible for all interactions with file systems.
```

File: packages/filesystem/src/browser/file-service.ts [hunk 5]
```diff
@@ -225,6 +259,12 @@ export class TextFileOperationError extends FileOperationError {
...
+ * This service manages all registered {@link FileSystemProvider}. In a nutshell, the {@link FileService} acts as a facade for all registered
+ * {@link FileSystemProvider}s and will forward calls to the responsible {@link FileSystemProvider} (this is determined by the scheme).
```
Comment (koegel) [inline packages/filesystem/src/browser/file-service.ts:265]:
```suggestion
 *  forwards calls to the responsible {@link FileSystemProvider}, determined by the scheme.
```

**Review by koegel — CHANGES_REQUESTED** [review]
Some minor suggestions and a question

File: packages/filesystem/src/browser/file-service.ts (hunk 2 above)
Comment (tortmayr) [reply packages/filesystem/src/browser/file-service.ts:219]: I don't know. I cannot derive this from the given code. Every time this event is fired the `provider` property is defined. So in my opinion this should not be optional.

File: packages/filesystem/src/browser/file-service.ts (hunk 1 above)
Comment (tortmayr) [reply packages/filesystem/src/browser/file-service.ts]: Fixed

File: packages/filesystem/src/browser/file-service.ts (hunk 2 above)
Comment (koegel) [reply packages/filesystem/src/browser/file-service.ts:220]: OK, maybe this is something we should bring up to the Theia devs on a different channel, we will discuss

**Review by koegel — APPROVED** [review]

File: packages/filesystem/src/common/files.ts [hunk 6]
```diff
@@ -558,85 +558,291 @@ export function ensureFileSystemProviderError(error?: Error): Error {
...
+    /** Event that is fired if an error occurred when watching files in the filesystem of this provider. */
     readonly onFileWatchError: Event<void>;
+
+    /**
+     * Watch the given resource and react to changes by firing the {@link FileSystemProvider#onDidChangeFile} event.
+     * @param resource `URI` of the resource to be watched.
+     * @param opts Options to define if the resource should be watched recursivly and to
+     *  provide a set of resources that should be excluded from watching.
+     *
+     * @retunrs A `Disposable` that can be invoked to stop watching the resource.
```
Comment (tortmayr) [inline packages/filesystem/src/common/files.ts:587]:
```suggestion
     * @returns A `Disposable` that can be invoked to stop watching the resource.
```

**Review by koegel — APPROVED** [review]

**Comment by sdirix** [comment]
Merged upstream with https://github.com/eclipse-theia/theia/pull/8596

**Pull request closed without merging**